composer/datasets/streaming/download.py

```diff
@@ -96,5 +126,6 @@ def download_or_wait(remote: str, local: str, wait: bool = False, max_retries: i
             error_msgs.append(e)
             last_error = e
             continue
-    if last_error:
-        raise RuntimeError(f"Failed to download {remote} -> {local}. Got errors:\n{error_msgs}") from last_error
+    if len(error_msgs):
+        print (f"Failed to download {remote} -> {local}. Got errors:\n{error_msgs}")
```

Comment:
Can you leave this line as it was? Would prefer to re-raise the exception. This looks like debugging code?

Reply:
We've got an issue with child threads not killing the parent dataloader thread... so I added the `os._exit(1)` line which should work, but then I'm not sure if I can raise an Error right before calling that... do you know of a better way?

Reply:
So the trouble I've been seeing is that exceptions raised in the child download threads are not killing the parent thread (a dataloader CPU worker). I think this is maybe expected, as `daemon=True` means that parent exit will cause child exit, but does not guarantee vice-versa.

My multithreading knowledge is very limited btw so it may be that there's a simple solution I'm not seeing.

Reply:
Here's where the child thread is created: https://github.com/mosaicml/composer/blob/dfc5a436df9fe2d045ef7ac52e314f728655173a/composer/datasets/streaming/dataset.py#L329

Reply:
Ah yes. So, the main thread should periodically check to see if the child thread died. If it did, then it should exit. See an example here:

https://github.com/mosaicml/composer/blob/33acb2525c811fcb5cfdaba5c35d0f0c421b16ea/composer/loggers/object_store_logger.py#L267

Reply:
Reverted error handling to what it was doing before, and going to move this `_check_workers` refactor to a follow-up PR to avoid blocking.